### PR TITLE
`ConcurrentState`

### PR DESCRIPTION
**Pull request by prestwich**
Adds a concurrently accessible version of the `State<Db>` from Revm. This manages a cache that can be updated by multiple threads via dashmap and an rwlock. The transitions can NOT be shared and require `&mut self`. The idea is that we should `Arc` it and then drop the users and use `Arc::get_mut` to manage transitions

### CODE REVIEW

**Review by Evalir — APPROVED** [review]
So this seems sound to me, as `dashmap` does a lot of the heavy lifting here and allows us to `Arc` this easily, and we're concentrating on concurrent reads. Could you expand a bit more on how you plan to handle the transitions though? this is probably not relevant to this PR itself, but I want to understand what's your thinking around it

File: src/evm.rs [hunk 1]
```diff
@@ -883,9 +883,9 @@ impl<'a, Ext, Db: Database + DatabaseCommit, TrevmState: HasBlock> Trevm<'a, Ext
 impl<Ext, Db: Database> EvmNeedsBlock<'_, Ext, State<Db>> {
     /// Finish execution and return the outputs.
     ///
-    /// ## Panics
-    ///
-    /// If the State has not been built with StateBuilder::with_bundle_update.
+    /// If the State has not been built with
```
Comment (prestwich) [inline src/evm.rs:886]: this is a drive-by change. underlying revm behavior changed
